docs-chatbot/generate-embeddings.ts

```diff
@@ -0,0 +1,100 @@
+import { promises as fs } from "fs";
+import { join } from "path";
+import dotenv from "dotenv";
+import { encode } from "gpt-tokenizer";
+import * as edgedb from "edgedb";
+import e from "dbschema/edgeql-js";
+import { initOpenAIClient } from "./utils";
+
+dotenv.config({ path: ".env.local" });
+
+const openai = initOpenAIClient();
+
+interface Section {
+  id?: string;
+  path: string;
+  tokens: number;
+  content: string;
+  embedding: number[];
+}
+
+async function walk(dir: string): Promise<string[]> {
+  const immediateFiles = await fs.readdir(dir);
+
+  const recursiveFiles: string[][] = await Promise.all(
+    immediateFiles.map(async (file: any) => {
+      const path = join(dir, file);
+      const stats = await fs.stat(path);
+      if (stats.isDirectory()) return walk(path);
+      else if (stats.isFile()) return [path];
+      else return [];
+    })
+  );
+
+  const flattenedFiles: string[] = recursiveFiles.reduce(
+    (all, folderContents) => all.concat(folderContents),
+    []
+  );
+
+  return flattenedFiles.sort((a, b) => a.localeCompare(b));
+}
```

Comment:
It's not that important, but you can simplify this a bit with the `withFileType` option to `readdir`, and use `Array.flat()` instead of a reduce. (Also is sorting the files needed?)
```ts
async function walk(dir: string): Promise<string[]> {
  const entries = await fs.readdir(dir, { withFileTypes: true });
  return (
    await Promise.all(
      entries.map((entry) => {
        const path = join(dir, entry.name);
        if (entry.isFile()) return [path];
        else if (entry.isDirectory()) return walk(path);
        return [];
      })
    )
  ).flat();
}
```

Reply:
Or in nodejs v20, there's a `recursive` option, but we should probably stick to lts as the min version for this.